### ISSUE TITLE
Can we use wandb sweep with the wandb callbacks provided?

### ISSUE THREAD

**Issue by shrave** [clementchadebec/benchmark_VAE#37]
Is there a way to integrate wandb sweep with the available wandb callback available?
If not, could you tell me how to exactly catch the loss values of a model to integrate onto the wandb sweep?

**Comment by clementchadebec** [comment]
Hi @shrave,

Thank you for opening this issue. Yes it is possible to use `wandb` sweep with the provided callbacks since it automatically logs the loss values :) Say you want to prospect two different latent space dimensions for a `VAE` and want to use learning rates in [0.0001, 0.001]. All you have to do is define the usual `sweep.yamp` file as follows:
```yaml
program: train_sweep.py
project: test-project
entity: benchmark_team
method: bayes
metric:
  name: eval/epoch_loss 
  goal: minimize
parameters:
  learning_rate:
    min: 0.0001
    max: 0.001
  latent_dim:
    values: [4, 16]
```  

and create your `train_sweep.py` which is a simple training script using a argparser for instance. An example file may look like this (removing all the data loading stuff)

```python
import argparse

from pythae.models import VAE, VAEConfig
from pythae.trainers import BaseTrainerConfig
from pythae.trainers.training_callbacks import WandbCallback

ap = argparse.ArgumentParser()

# Training setting
ap.add_argument(
    "--learning_rate",
    default=1e-3,
    type=float
)
ap.add_argument(
    "--latent_dim",
    default=2,
    type=int
)

args = ap.parse_args()

model_config = VAEConfig(
    latent_dim=args.latent_dim # pass the sweep hyper-parameter 
)

model = VAE(
    model_config=model_config
)

training_config = BaseTrainerConfig(
    learning_rate = args.learning_rate, # pass the sweep hyper-parameter 
    num_epochs = args.max_epochs

)

callbacks = []

wandb_cb = WandbCallback()
wandb_cb.setup(
    training_config,
    model_config=model_config
)

callbacks.append(wandb_cb)

pipeline = TrainingPipeline(training_config=training_config, model=model)
pipeline(train_data=train_data, eval_data=eval_data, callbacks=callbacks)

if __name__ == "__main__":
    main(args)
```

Then you can login to wandb, setup the sweep and launch it as usual.

I hope this helps.

Clément

**Comment by shrave** [comment]
Hi,
Thank you very much! The script particularly was very helpful.
Thanks for your quick response.

**Comment by shrave** [comment]
Hi, 
Had another question, can we log quantities or custom metrics other than training/eval losses using the wandb callbacks in pythae?
If so, could you show how?

Thank you.

**Comment by clementchadebec** [comment]
Hi @shrave,

Sorry for the late reply. Yes, you can but you will have to amend the trainers instances a bit by yourself.  First, you will need to clone the repo and install it in editable mode
```bash
$ git clone https://github.com/clementchadebec/benchmark_VAE.git
$ cd benchmark_VAE
$ pip install -e .
```

Then, you can add your metrics to the `metrics` dict directly in the trainer as done below for the `train_epoch_loss`https://github.com/clementchadebec/benchmark_VAE/blob/0f5c0cc5aa361e5bea7294c02502028904a0305f/src/pythae/trainers/base_trainer/base_trainer.py#L332

The callback should then handle the metrics and log them automatically to `wandb`.
Please note that, by convention, train metrics are named `train_`, validation `val_` and test `test_`.

Do not hesitate if you have any questions :)

Best,

Clément
